src/_TextInputWrapper.tsx

```diff
@@ -25,45 +43,50 @@ type StyledWrapperProps = {
   hasIcon?: boolean
   block?: boolean
   contrast?: boolean
+  validationStatus?: 'error' | 'warning'
+  /** @deprecated Use `size` prop instead */
   variant?: 'small' | 'large'
+  size?: 'small' | 'large'
 } & WidthProps &
   MinWidthProps &
   MaxWidthProps &
   SxProp
 
 const TextInputWrapper = styled.span<StyledWrapperProps>`
-  display: inline-flex;
-  align-items: stretch;
+  width: max-content;
   min-height: 34px;
   font-size: ${get('fontSizes.1')};
   line-height: 20px;
   color: ${get('colors.fg.default')};
   vertical-align: middle;
+  background-color: ${get('colors.canvas.default')};
   background-repeat: no-repeat; // Repeat and position set for form states (success, error, etc)
   background-position: right 8px center; // For form validation. This keeps images 8px from right and centered vertically.
   border: 1px solid ${get('colors.border.default')};
   border-radius: ${get('radii.2')};
   outline: none;
   box-shadow: ${get('shadows.primer.shadow.inset')};
   cursor: text;
+  padding: 6px 12px;
+  display: grid;
+  grid-template-areas: 'leadingIcon input trailingIcon';
```

Comment:
We should add `grid-template-columns: auto 1fr auto;` so that the empty grid cells collapse when no content is passed.

Stuff like this will happen if we don't:
![Screen Shot 2021-11-30 at 3 29 10 PM](https://user-images.githubusercontent.com/2313998/144123540-5fd133d3-5bf8-4af3-be0e-1ed16b9d738d.png)

Reply:
Yeah, I've been trying to fix this as well.
But your solution results in 
![image](https://user-images.githubusercontent.com/417268/144197315-1af194a5-8af9-416a-8b5e-b7b42e75c743.png)

Reply:
Adding 
```
grid-template-columns: auto 1fr auto;
justify-items: end;
```
seems to help. WDYT?

Reply:
It looks like the text input is getting sized with `auto` when it's the first child. We might need to change the `grid-template-columns` based on what children are there.

#### `leadingVisual` + input + `trailingVisual`
`grid-template-columns: auto 1fr auto;`

#### `leadingVisual` + input
`grid-template-columns: auto 1fr;`

#### input + `trailingVisual`
`grid-template-columns: 1fr auto;`

#### input
`grid-template-columns: 1fr;`
or
`grid-template-columns: unset;`

Reply:
I was hoping grid would have some fancy way of handling this without having to specify different column templates for each case, but I couldn't figure it out :(

Maybe another reviewer will know?

Reply:
Or maybe we could just use flexbox instead of grid? Is there something we need from grid that can't be done with flexbox?

Reply:
Yeah I'm not particularly attached to the grid way. Do you want to pair on this for a bit tomorrow?

Reply:
Sounds great! I love a CSS pairing sesh :)